### PR TITLE
[ITV] Update ITV extractor from yt-dlp back-port

### LINKED ISSUE #25796: Unable to download subtitles
YouTube-dl failed to download the subtitles for this program that it had found.
I've managed to download subs for other programmes that had them.

V:\YouTube-dl>youtube-dl --all-subs --skip-download  https://www.itv.com/hub/agatha-christies-marple/9Da13066
[ITV] 9Da13066: Downloading webpage
[ITV] 9Da13066: Downloading XML
WARNING: Unable to download XML: HTTP Error 504: Gateway Time-out
[ITV] 9Da13066: Downloading JSON metadata
[ITV] 9Da13066: Downloading m3u8 information
[info] Writing video subtitles to: The Body in the Library-9Da13066.en.vtt
WARNING: Unable to download subtitle for "en": Unable to download webpage: HTTP Error 404: Not Found (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.

V:\YouTube-dl>youtube-dl -v
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v']
[debug] Encodings: locale cp1252, fs mbcs, out cp850, pref cp1252
[debug] youtube-dl version 2020.06.16.1
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.18362
[debug] exe versions: ffmpeg git-2020-03-15-c467328, ffprobe git-2020-03-15-c467328
[debug] Proxy map: {}
Usage: youtube-dl [OPTIONS] URL [URL...]

### LINKED ISSUE #27582: Unable to download from itv.com
## Checklist

- [x ] I'm reporting a broken site support
- [ x] I've verified that I'm running youtube-dl version **2020.12.29**
- [x ] I've checked that all provided URLs are alive and playable in a browser
- [x ] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x ] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
PASTE VERBOSE LOG HERE
```
[debug] System config: []
[debug] User config: ['-o', 'F:/8Tb-Videos/_For-Processing/%(title)s.%(ext)s', '--ffmpeg-location', 'C:/Progra~2/get_iplayer/utils']
[debug] Custom config: []
[debug] Command-line args: ['-v', 'https://www.itv.com/hub/bugsy-malone/CFD0085']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2020.12.29
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.19041
[debug] exe versions: ffmpeg 4.2.1
[debug] Proxy map: {}
[debug] Using fake IP 25.181.70.25 (GB) as X-Forwarded-For.
[ITV] CFD0085: Downloading webpage
[ITV] CFD0085: Downloading JSON metadata
[ITV] CFD0085: Downloading m3u8 information
WARNING: Failed to download m3u8 information: HTTP Error 403: Forbidden
ERROR: No video formats found; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmp581_vk6o\build\youtube_dl\YoutubeDL.py", line 803, in wrapper
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmp581_vk6o\build\youtube_dl\YoutubeDL.py", line 824, in __extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmp581_vk6o\build\youtube_dl\extractor\common.py", line 532, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmp581_vk6o\build\youtube_dl\extractor\itv.py", line 115, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmp581_vk6o\build\youtube_dl\extractor\common.py", line 1367, in _sort_formats
youtube_dl.utils.ExtractorError: No video formats found; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.

## Description

WRITE DESCRIPTION HERE

I had a similar problem a few days ago, but my version was a week or two out of date. Updating to the latest version fixed it until today.

### LINKED ISSUE #28906: ITV Hub JSON error
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2021.04.26**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['https://www.itv.com/hub/tipping-point/2a1875a1123', '--verbose']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2021.04.26
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.19041
[debug] exe versions: none
[debug] Proxy map: {}
[debug] Using fake IP 25.161.160.85 (GB) as X-Forwarded-For.
[ITV] 2a1875a1123: Downloading webpage
[ITV] 2a1875a1123: Downloading JSON metadata
ERROR: Unable to download JSON metadata: HTTP Error 500: Internal Server Error (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpeav8wi2f\build\youtube_dl\extractor\common.py", line 634, in _request_webpage
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpeav8wi2f\build\youtube_dl\YoutubeDL.py", line 2279, in urlopen
  File "C:\Python\Python34\lib\urllib\request.py", line 470, in open
  File "C:\Python\Python34\lib\urllib\request.py", line 580, in http_response
  File "C:\Python\Python34\lib\urllib\request.py", line 508, in error
  File "C:\Python\Python34\lib\urllib\request.py", line 442, in _call_chain
  File "C:\Python\Python34\lib\urllib\request.py", line 588, in http_error_default
```

## Description

Several programmes on ITV Hub now fail to download. URL's such as

https://www.itv.com/hub/the-chase/1a7842a1761 
https://www.itv.com/hub/tipping-point/2a1875a1123
https://www.itv.com/hub/tenable/10a0805a0021

fail to download though some like 

https://www.itv.com/hub/lorraine/1a9360a2830

continue to work

### LINKED ISSUE #29620: youtube-dl https://www.itv.com/hub/professor-t/7a0171a0001 fails with 404 error
This is the --verbose output:

ERROR: Unable to download JSON metadata: HTTP Error 404: Not Found (caused by <HTTPError 404: 'Not Found'>); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/usr/lib/python3.9/site-packages/youtube_dl/extractor/common.py", line 634, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/lib/python3.9/site-packages/youtube_dl/YoutubeDL.py", line 2287, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib64/python3.9/urllib/request.py", line 523, in open
    response = meth(req, response)
  File "/usr/lib64/python3.9/urllib/request.py", line 632, in http_response
    response = self.parent.error(
  File "/usr/lib64/python3.9/urllib/request.py", line 561, in error
    return self._call_chain(*args)
  File "/usr/lib64/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/usr/lib64/python3.9/urllib/request.py", line 641, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)

Version in use is:
youtube-dl-2021.06.06-1.fc34.noarch

Thanks.

### LINKED ISSUE #30265: Can't download ITV hub show
## Checklist

- [x] I'm asking a question
- [x] I've looked through the README and FAQ for similar questions
- [x] I've searched the bugtracker for similar questions including closed ones

## Question

I am trying to download a show from the ITV hub (here: https://www.itv.com/hub/the-ellen-degeneres-show/2a4350a1024). Using Stream Detector on Firefox, I get the following code to paste into Youtube-dl:

`youtube-dl --no-part --restrict-filenames --user-agent "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:94.0) Gecko/20100101 Firefox/94.0" --add-header "Cookie:Itv.Session={%22tokens%22:{%22content%22:{%22entitlement%22:{%22purchased%22:[]%2C%22failed_availability_checks%22:[]%2C%22source%22:%22%22}%2C%22email_verified%22:true%2C%22access_token%22:%22eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJodHRwczovL2F1dGguaXR2LmNvbSIsInN1YiI6IjZkYmFmYWU3LTU0NzYtNDNmYS1hYWEwLTAwODQxYmY2OTFkMSIsImV4cCI6MTYzNzg0MjMwMiwiaWF0IjoxNjM3NzUyMzAyLCJicm9hZGNhc3RFcnJvck1zZyI6IiIsImJyb2FkY2FzdFJlc3BvbnNlQ29kZSI6IjIwMCIsImJyb2FkY2FzdGVyIjoiSVRWIiwiaXNBY3RpdmUiOnRydWUsIm5vbmNlIjoiZGdSQ1gzWWVkNURvWWl6VjhQN20iLCJuYW1lIjoiU2VhbiIsInNjb3BlIjoiY29udGVudCIsImVudGl0bGVtZW50cyI6W10sInBheW1lbnRTb3VyY2UiOiIiLCJzaG93UHJpdmFjeU5vdGljZSI6ZmFsc2UsInVuZGVyMTgiOmZhbHNlfQ.Da_C_X8gHGWh6QBo1Im_hqspwvo9z8XVKkItpCqKW9BYtJVivSSB5eqO-b1Q9U8fk9QN5IqXTM2nOHe22BUTm02rcKUXgsTJ41pRqaAuAquuSXkPpZSnOgbiE0flsEnQN78mfTzs0k9jvLO-A0QT939NowHCcd8_BfbdFCrE11FI0zCFyTHKatVuS9j1bDK98Nd2O61iphtBcmNBwaql1p94okb6QWYSI3kqFyNBICgziTwi-6t-kCZ76_wSopIPUuQ21oRz4911esS3kBId4SdrUcSpezAgUS-_GS3XpCMiOLWGJ69fXU3qzkXP8NRLYSxcf_RTU4QkmWs6kjvaXg%22%2C%22token_type%22:%22bearer%22%2C%22refresh_token%22:%22eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJodHRwczovL2F1dGguaXR2LmNvbSIsInN1YiI6IjZkYmFmYWU3LTU0NzYtNDNmYS1hYWEwLTAwODQxYmY2OTFkMSIsImV4cCI6NDc5MzQyNTkwMiwiaWF0IjoxNjM3NzUyMzAyLCJub25jZSI6ImRnUkNYM1llZDVEb1lpelY4UDdtIiwic2NvcGUiOiJjb250ZW50IiwiYXV0aF90aW1lIjoxNjM3NzUyMzAyfQ.GV1cqako7yaGvA393O9oa9cVhX0I-Sw5M2wSR1gg5RAYt1lzYIY0T2nLUu3xLte8p1JqlAVb_1IEV_Yts12dxQTHVmDupqA33dKIlA0Pxxkq1Mv8x5WjjOuVgnmjf1lmm7Ygt8560V9olis5fKlONa4QNrsX1c4O0RNr630WzPBbFk6e1PsGJ2bVOBxp5ja8kzun-_oOCjonrQOxTHj3-e9RmyCVJdbRLzpFkfJf_n03zNXly_d-uBad3vodfHemv8dQJeXXKSDQGVAq3idlPj77bEHfRkfmE6yUMBjMlZSF_C3HfqomFyN84wNw9JiQLMumwP2tHNpfnYVvqBWy4Q%22}}%2C%22sticky%22:true}; mid=1709081696670734686; _cs_c=1; _cs_id=8fcd982a-e516-ad59-e9f0-5ec3cfcfacdc.1637751312.1.1637752301.1637751312.1.1671915312583; _cs_s=7.0.0.1637754101993; __utma=187186800.658762206.1637751313.1637751313.1637751313.1; __utmb=187186800.13.9.1637752166735; __utmc=187186800; __utmz=187186800.1637751313.1.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none); _hjSessionUser_1586194=eyJpZCI6IjcyMmJhMjBmLWU5MzEtNTA1YS1iYjYyLWFhMTI3NTBkMTViOSIsImNyZWF0ZWQiOjE2Mzc3NTEzMTMwODcsImV4aXN0aW5nIjp0cnVlfQ==; _hjFirstSeen=1; _hjSession_1586194=eyJpZCI6IjQ2MTY2NDM1LThhZTQtNGJhOC05NjAzLTM4ZTBjOGVjMjY1MiIsImNyZWF0ZWQiOjE2Mzc3NTEzMTMwOTN9; _hjAbsoluteSessionInProgress=0; Itv.CookiePolicy.v2=accepted; Itv.Region=ITV|null; __utmt_UA-17825253-53=1" --referer "https://www.itv.com/" "https://itvpnpdotcom.cdn1.content.itv.com/2-4350-1024-001/18/1/VAR028/2-4350-1024-001_18_1_VAR028.ism/.mpd?Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cHM6Ly9pdHZwbnBkb3Rjb20uY2RuMS5jb250ZW50Lml0di5jb20vMi00MzUwLTEwMjQtMDAxLzE4LzEvVkFSMDI4LzItNDM1MC0xMDI0LTAwMV8xOF8xX1ZBUjAyOC5pc20vKiIsIkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTYzNzc3NTcwM319fV19&Signature=InPnv~NiUQzV2St09QxkG5u63iiUWKbWsV4V5csiShiz0YOEVf9gy7OvYAPRCPXjSg2T95wzshNY8Whl92I2AMBv69jyyrY0s8zGI7ja6dMxZB0mbxyg5L8M15Y2TnwwgTSnfqpDfX9HkZdBZXnhRt~Ko6cbPf-Ao8NcI0-DcNFKrAuw8AQIECo-F4YFdECk9AcqPcgWvIVJngUYhQmQY97foz2~13UhTK7H-DbGqyYyocgt-YDeNfBCnLfoBotVDqe2q6laK81wX5BGHiEKzj44t3hHqE-H4GUhfq8vHeHYbyeR92pDSjAlNGuW4CYxQqXRTQl70fI6dN~PP32rKg__&Key-Pair-Id=APKAJB7PCFZAZHWZVIBA"`

However, this produces the following error: 

`ERROR: No video formats found; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.`

### LINKED ISSUE #30578: Unable to download from ITV Hub
## Checklist

- [x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar bug reports including closed ones
- [x] I've read bugs section in FAQ

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'https://www.itv.com/hub/celebrity-juice/1a7045a0288']
[debug] Encodings: locale UTF-8, fs UTF-8, out None, pref UTF-8
[debug] youtube-dl version 2021.12.17
[debug] Python version 2.7.18 (CPython) - Linux-5.4.0-96-generic-x86_64-with-Ubuntu-20.04-focal
[debug] exe versions: ffmpeg 4.2.4, ffprobe 4.2.4, phantomjs ., rtmpdump 2.4
[debug] Proxy map: {'no': 'localhost,127.0.0.1,::1'}
[debug] Using fake IP 25.170.173.6 (GB) as X-Forwarded-For.
[ITV] 1a7045a0288: Downloading webpage
ERROR: Unable to download webpage: ('The read operation timed out',) (caused by SSLError('The read operation timed out',)); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 634, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 2288, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/usr/lib/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/local/bin/youtube-dl/youtube_dl/utils.py", line 2737, in https_open
    req, **kwargs)
  File "/usr/lib/python2.7/urllib2.py", line 1208, in do_open
    r = h.getresponse(buffering=True)
  File "/usr/lib/python2.7/httplib.py", line 1178, in getresponse
    response.begin()
  File "/usr/lib/python2.7/httplib.py", line 452, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python2.7/httplib.py", line 408, in _read_status
    line = self.fp.readline(_MAXLINE + 1)
  File "/usr/lib/python2.7/socket.py", line 480, in readline
    data = self._sock.recv(self._rbufsize)
  File "/usr/lib/python2.7/ssl.py", line 754, in recv
    return self.read(buflen)
  File "/usr/lib/python2.7/ssl.py", line 641, in read
    v = self._sslobj.read(len)
```

## Description

Trying to download episodes of Celebrity Juice from the ITV Hub, but it's halting after 'Downloading webpage' and then several minutes later it errors out.

### LINKED ISSUE #30994: [ITV] SSLError: read timeout (from #30992)
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'-f', u'bestvideo+bestaudio/best', u'-o', u'%(title)s.%(ext)s', u'-ciw', u'--write-sub', u'--convert-subs', u'srt', u'https://www.itv.com/hub/isle-of-man-tt/1a7918a0188https://www.itv.com/hub/isle-of-man-tt/1a7918a0188']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2021.12.17
[debug] Python version 2.7.18 (CPython) - Linux-4.4.0-18362-Microsoft-x86_64-with-Ubuntu-20.04-focal
[debug] exe versions: ffmpeg 4.2.4, ffprobe 4.2.4
[debug] Proxy map: {}
[debug] Using fake IP 25.80.8.219 (GB) as X-Forwarded-For.
[ITV] 1a7918a0188https: Downloading webpage

ERROR: Unable to download webpage: ('The read operation timed out',) (caused by SSLError('The read operation timed out',)); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 634, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 2288, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/usr/lib/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/local/bin/youtube-dl/youtube_dl/utils.py", line 2737, in https_open
    req, **kwargs)
  File "/usr/lib/python2.7/urllib2.py", line 1208, in do_open
    r = h.getresponse(buffering=True)
  File "/usr/lib/python2.7/httplib.py", line 1178, in getresponse
    response.begin()
  File "/usr/lib/python2.7/httplib.py", line 452, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python2.7/httplib.py", line 408, in _read_status
    line = self.fp.readline(_MAXLINE + 1)
  File "/usr/lib/python2.7/socket.py", line 480, in readline
    data = self._sock.recv(self._rbufsize)
  File "/usr/lib/python2.7/ssl.py", line 754, in recv
    return self.read(buflen)
  File "/usr/lib/python2.7/ssl.py", line 641, in read
    v = self._sslobj.read(len)
```

## Description

Was trying to download ITV Hub programme, unsure if this is support as I have used ytdl before for ITV content.

### LINKED ISSUE #31363: itvhub download stopped working
When I try to download from itvhub now I get this response
```
$ yt-dlp https://www.itv.com/watch/the-amazing-spider-man-2/2a3878
[generic] 2a3878: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] 2a3878: Extracting information
ERROR: Unsupported URL: https://www.itv.com/watch/the-amazing-spider-man-2/2a3878
```
everything was fine until today.

Im running MX-Linux and yt-dlp version
```
$ yt-dlp -U
Latest version: 2022.11.11, Current version: 2022.11.11
yt-dlp is up to date (2022.11.11)
```
TIA

Brian

### PR DESCRIPTION
**Pull request by dirkf**
## Please follow the guide below

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/), except where
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (`yt-dlp/yt_dlp/extractor/itv.py`)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This PR back-ports the [ITV extractor from yt-dlp](https://github.com/yt-dlp/yt-dlp/pull/1776), so incorporating all the latest fixes and facilitating future joint updates.

The PR from #28955 is included for the record as an an interim commit.

In addition the following changes are applied:
* work-around for ITV using `:` as a decimal point in duration value
* set default UA for extraction as `Mozilla/5.0` to avoid read time-out in extraction
* ditto for download to avoid 403
* detect and handle 404 error for `Episode not found` and `We're really sorry`.

Resolves #25796.
Resolves #27582.
Resolves #28906.
Resolves #29620.
Resolves #30265.
Resolves #30578.
Resolves #30994.

Closes #28955.

Now updated for ITVX instead/as well.

Resolves #31363.

Also updated ITV BTCC, handles news articles.

Supersedes, closes #25254.

### CODE REVIEW

**Comment by stsquad** [comment]
I just want to confirm that testing this branch I was able to download what I was looking to where the current upstream master would just timeout and fail.

**Comment by dirkf** [comment]
Thanks. It's probably not worth merging this just now since ITV is about to revamp its streaming/catchup services.

File: youtube_dl/extractor/itv.py [hunk 1]
```diff
@@ -10,33 +10,29 @@
...
     extract_attributes,
-    get_element_by_class,
-    JSON_LD_RE,
+    get_element_by_attribute,
+    int_or_none,
     merge_dicts,
     parse_duration,
+    parse_iso8601,
     smuggle_url,
+    std_headers,
```
Comment (dirkf) [inline youtube_dl/extractor/itv.py:19]:
```suggestion
```

File: youtube_dl/extractor/itv.py [hunk 2]
```diff
@@ -49,109 +122,199 @@ class ITVIE(InfoExtractor):
...
+            'channel': video_str('channel'),
+            'categories': traverse_obj(video_data, ('categories', 'formatted'), expected_type=list),
+            'age_limit': {False: 16, True: 0}.get(video_data.get('isChildrenCategory')),
         }, info)
 
 
-class ITVBTCCIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?itv\.com/btcc/(?:[^/]+/)*(?P<id>[^/?#&]+)'
+class ITVBTCCIE(ITVBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?itv\.com/(?!:watch/|hub/)(?:[^/]+/)+(?P<id>[^/?#&]+)'
```
Comment (dirkf) [inline youtube_dl/extractor/itv.py:317]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?itv\.com/(?!(?:watch|hub)/)(?:[^/]+/)+(?P<id>[^/?#&]+)'
```

**Comment by MysticMerlin** [comment]
Hi,
I'm not a complete novice but I'm having some difficulty following what's going on & whether the problem with the ITVX has been fixed. I can see that dirkf has been very busy & I'h sure there are many thousands of people appreciating all his efforts & anybody else that is capable of contributing. I'm afraid I don't have the experience or knowledge to offer any contribution from my side. So, in layman's terms where are we at as regards ITVX? Thanks.

**Comment by dirkf** [comment]
The updated extractor is now in the git master repository code that would be used to make a release. A release built from git master should support ITVX. Also, installing from the git master with _pip_ should do so.

**Comment by psl3** [comment]
I have not managed to get Python to work (or understand it) - I'm guessing it is just a wait until a kind and skilled person does a release with an exe file that will support ITVX (watching both youtube-dl and yt-dlp for an updated file).
